File: linked_lists/linked_lists.js

// Singly-linked List
var Node = function(data) {

  this.next = null;
  this.data = data;

  this.appendToTail = function(data) {
    var n = this;
    while (n.next != null) {
      n = n.next;
    };
    n.next = new Node(data);
  };
  
};

var LinkedList = function(node) {
  this.head = node;

  this.appendToTail = function(data) {
    this.head.appendToTail(data);
  };

  this.length = function() {
    var i = 1;
    var node = this.head;
    while (node.next !== null) {
      node = node.next;
      i++;
    };
    return i;
  }

  this.deleteNodeByValue = function(value) {
    var n = this.head;

    if (n.data === value) {
      this.head = n.next;
      console.log(this.head)
      return;
    };

    while (n !== null) {
      if (n.next.data === value) {
        n.next = n.next.next;
        break;
      };
      n = n.next;
    };

    return;

  };

};

a = new Node('a');
a.appendToTail('b');

list = new LinkedList(a);
list.appendToTail('c');

list.deleteNodeByValue('a');

console.log(list.length());

File: linked_lists/linked_lists.rb
class Node
  attr_accessor :next
  attr_reader :datum
  
  def initialize(datum)
    @next = nil
    @datum = datum
  end

  def append_to_tail(datum)
    current_node = self
    while (current_node.next != nil)
      current_node = current_node.next
    end
    current_node.next = Node.new(datum)
  end

end

class LinkedList
  attr_reader :head, :length

  def initialize(node)
    @head = node
  end

  def append_to_tail(datum)
    self.head.append_to_tail(datum)
  end

  def length
    current_node = self.head
    i = 0
    while (current_node != nil)
      current_node = current_node.next
      i += 1
    end
    i
  end
end

# a = Node.new('a')

# list = LinkedList.new(a)
# list.append_to_tail('b')
# list.append_to_tail('c')
# list.append_to_tail('d')

# puts list.length

class DoublyLinkedNode
  attr_accessor :next, :prev
  attr_reader :datum
  
  def initialize(datum)
    @next = nil
    @prev = nil
    @datum = datum
  end

  def append_to_tail(datum)
    current_node = self
    while (current_node.next != nil)
      current_node = current_node.next
    end
    current_node.next = DoublyLinkedNode.new(datum)
    current_node.next.prev = self
  end

end

a = DoublyLinkedNode.new('a')
list = LinkedList.new(a)

puts list.length
list.append_to_tail('b')
list.append_to_tail('c')
puts list.length